src/twisted/python/htmlizer.py

```diff
@@ -60,35 +71,58 @@ def __init__(self, writer):
         reflect.accumulateClassList(self.__class__, "noSpan", noSpan)
         self.noSpan = noSpan
 
+
     def write(self, token, type=None):
-        token = cgi.escape(token)
+        if _PY3 and isinstance(token, bytes):
+            token = token.decode("utf-8")
+        token = escape(token)
+        if _PY3:
+            token = token.encode("utf-8")
         if (type is None) or (type in self.noSpan):
             self.writer(token)
         else:
-            self.writer('<span class="py-src-%s">%s</span>' %
-                        (type, token))
+            self.writer(
+                b'<span class="py-src-' + type.encode("utf-8") + b'">' +
+                token + b'</span>')
+
 
 
 class SmallerHTMLWriter(HTMLWriter):
-    """HTMLWriter that doesn't generate spans for some junk.
+    """
+    HTMLWriter that doesn't generate spans for some junk.
 
     Results in much smaller HTML output.
     """
     noSpan = ["endmarker", "indent", "dedent", "op", "newline", "nl"]
 
+
+
 def filter(inp, out, writer=HTMLWriter):
-    out.write('<pre>')
+    if _PY3:
+        tokenizeFunc = tokenize.tokenize
+    else:
+        tokenizeFunc = tokenize.generate_tokens
+
+    out.write(b'<pre>')
     printer = TokenPrinter(writer(out.write).write).printtoken
     try:
-        tokenize.tokenize(inp.readline, printer)
+        for token in tokenizeFunc(inp.readline):
+            (tokenType, string, start, end, line) = token
+            printer(tokenType, string, start, end, line)
     except tokenize.TokenError:
         pass
-    out.write('</pre>\n')
+    out.write(b'</pre>\n')
+
+
 
 def main():
     import sys
-    with open(sys.argv[1]) as f:
-        filter(f, sys.stdout)
+    if _PY3:
```

Comment:
The idiom here ought to be `getattr(sys.stdout, "buffer", sys.stdout)`, since not all instances of `sys.stdout` will have a buffer, even on py3.